docs/developing/local-development.markdown

```diff
@@ -9,20 +9,19 @@ If you want to develop Jellyfish on your local machine, the pre-requisites are:
 Install the dependencies:
 
 ```sh
-$ npm i
+npm i
 ```
 
 You can then run these commands in different terminal emulators, which will run
 all services in non-daemon mode:
 
 ```sh
-$ npm run compose:database
-$ SERVER_HOST=http://localhost SERVER_PORT=8000 UI_PORT=9000 npm run dev:ui
-$ SERVER_HOST=http://localhost SERVER_PORT=8000 POSTGRES_HOST=localhost REDIS_HOST=localhost npm run dev:server
+npm run compose:database
+SERVER_HOST=http://localhost SERVER_PORT=8000 UI_PORT=9000 npm run dev:ui
+SERVER_HOST=http://localhost SERVER_PORT=8000 POSTGRES_HOST=localhost METRICS_PORT=9100 SOCKET_METRICS_PORT=9101 REDIS_HOST=localhost npm run dev:server
```

Comment:
Things seem to run fine for me without setting `METRICS_PORT` and `SOCKET_METRICS_PORT`. Do you get an error?

Reply:
I set `METRICS_PORT` to 9100 because the default of 9000 collides with the port used for the UI on the previous sentence.
In my machine ( macOS ) both containers start. But the metrics server binds to IPv6 while JF UI binds to ipv4( see below for more details ). If I try to access JF I get a 404 because the browser connected to localhost ipv6, so to the metrics server.
```
Request URL: http://localhost:9000/
Request Method: GET
Status Code: 404 Not Found
Remote Address: [::1]:9000
Referrer Policy: strict-origin-when-cross-origin
```

We can check with `lsof`

```
╰─➤  lsof -i -n -P | grep 9000
node      26771 ramiro   24u  IPv4 0x38b072b6f95bc461      0t0  TCP *:9000 (LISTEN)
node      27020 ramiro   44u  IPv6 0x38b072bbc73b3029      0t0  TCP *:9000 (LISTEN)
```

the first one ( UI ) is bound to v4, and the second one ( dev:server ) to v6.

```
╰─➤  ps -el | grep 26771
  501 26771 26763     4006   0  31  0 67122468  40612 -      S+                  0 ttys007    0:32.98 webpack
  501 26790 26771     4006   0  31  0 34857904   8408 -      S+                  0 ttys007    0:04.16 /Users/ramiro/palomonte/github/product-os/jellyfish/apps/ui/node_modules/esbuild-darwin-64/bin/esbuild --service=0.14.23 --ping

╰─➤  ps -el | grep 27020
  501 27020 27018     4006   0  31  0 45968676 999348 -      S+                  0 ttys008    0:23.98 node -r ts-node/register --inspect=0.0.0.0 ./lib/index.ts
```

I also changed `SOCKET_METRICS_PORT=9101` to keep it consistent with METRICS_PORT, so that both use the same prefix of 910x.

Reply:
I've just tested on an Ubuntu machine and the behavior is different here. The dev server detects the port is in use and switches to a random one:

```
2022-04-11T13:55:38.918Z [jellyfish-metrics/build/index] [info] [SERVER-48.11.19-localhost-2ed107c3-6ff2-4cef-b344-e0410d544195]: Port 9000 is in use, starting metrics server on a random port
2022-04-11T13:55:38.919Z [jellyfish-metrics/build/index] [info] [SERVER-48.11.19-localhost-2ed107c3-6ff2-4cef-b344-e0410d544195]: Metrics server listening on port [object Object]
```

Reply:
@ramirogm Thanks for looking into this. It seems like it might be a good idea to set the ports manually (like you've done in this PR) just in case, but now I'm wondering why we don't see the same random port log on Mac. That is a part of the server bootstrap code, so I'm a bit confused. I develop on both Linux and WSL2 and do see these logs, but I also never use Prometheus/Grafana when running locally.

Reply:
As an alternative, I've made a change on `jellyfish-metrics` to bind only to IPv4 and also show the actual listening address/port instead of [object Object].

I'll link the PR here


```
2022-04-11T14:22:06.006Z [build/index] [info] [SERVER-48.11.10-localhost-a4914deb-d8fe-4719-82cb-a73e1810c5ae]: Starting metrics server on 9000
2022-04-11T14:22:06.010Z [build/index] [info] [SERVER-48.11.10-localhost-a4914deb-d8fe-4719-82cb-a73e1810c5ae]: Port 9000 is in use, starting metrics server on a random port
2022-04-11T14:22:06.011Z [build/index] [info] [SERVER-48.11.10-localhost-a4914deb-d8fe-4719-82cb-a73e1810c5ae]: Metrics server listening on port {"address":"0.0.0.0","family":"IPv4","port":56948}
```

Reply:
@ramirogm Thanks for looking into this a bit more, I've approved your metrics PR as I'd rather fix our logic/code instead of forcing developers to declare even more env vars. :+1: